### PR TITLE
Author Sorting

### PR DESCRIPTION
**Pull request by amohanan-scottlogic**
#Author Sorting

- Sorting by author name in ascending order

- test cases for the same

### CODE REVIEW

File: src/main/java/com/scottlogic/UserPost.java [hunk 1]
```diff
@@ -4,8 +4,10 @@
 import java.time.format.DateTimeFormatter;
 import java.util.Objects;
 
+import javax.validation.constraints.NotNull;
+
 public class UserPost {
-    private String author;
+    @NotNull private String author;
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/UserPost.java:10]: I'm not sure if this is really the appropriate use case for `@NotNull`. Without running it through a validator, it's not actually guarding against creating a UserPost with a null author - it just throws an exception if we try putting it through `Collections.sort()`.

We can look a bit deeper into it, but maybe leave it out for now and update your sorter and tests on the assumption that any field could be null.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            OffsetDateTime.of(2020, 3, 12, 13, 22, 12, 0, ZoneOffset.UTC),
+            "An example of a post \nwith lines breaks.", 3);
+
+    private UserPost userPost4 = new UserPost("May Bloggs",
+            OffsetDateTime.of(2020, 1, 3, 8, 53, 34, 0, ZoneOffset.UTC),
+            "Another example post.", 1);
+
+    private UserPost userPost5 = new UserPost(null,
+            OffsetDateTime.of(2020, 3, 12, 13, 22, 12, 0, ZoneOffset.UTC),
+            "An example of a post \nwith lines breaks.", 3);
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:35]: There's a lot of boilerplate code here for creating very similar objects. We're only really bothered about the author, so we can create a private helper method to return a UserPost with a specified author. It would then be better to call this from the individual test methods, so a reader can immediately see what that particular test is doing.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        List<UserPost> sorted = Arrays.asList(userPost1);
+        Assert.assertEquals(userPosts, sorted);
+    }
+
+    @Test
+    public void authorSorter_nullNames_returnNull() {
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4, userPost5);
+        List<UserPost> sorted = aps.authorSorter(userPosts);
+        Assert.assertNull(sorted);
+    }
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:64]: Try and think up some more test cases here. Think about edge cases, such as duplicated names, different flavours of "nothingness", etc.

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]
Great start! I may have led you astray with that `@NotNull` though...

File: src/main/java/com/scottlogic/PostSorter.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.scottlogic;
+
+import java.util.List;
+
+public interface PostSorter {
+    //List<UserPost> sort(List<UserPost> inputList);
+    List<UserPost> authorSorter(List<UserPost> inputList);
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/PostSorter.java:7]: Spent a lot of time on it this morning :). I've got the hang of it now. So now i've name the function sorter. Didn't realise I had commented out what was already there. I think I deleted it at some point

File: src/main/java/com/scottlogic/UserPost.java (hunk 1 above)
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/UserPost.java:10]: Currently, if any names are null, while sorting I return null and display "Sorting cannot be done".
I'm currently writing test cases for date sorting, and that does not allow me to enter any non integer values in the input

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java (hunk 2 above)
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/AuthorPostSorterTest.java:35]: Will look at implementing this today

File: src/main/java/com/scottlogic/UserPost.java (hunk 1 above)
Comment (pling-scottlogic) [reply src/main/java/com/scottlogic/UserPost.java:10]: I think I would still expect to get a sorted list back, even if I provide it with some null authors. I'd probably expect all the null authors to be at the end of the list.

File: src/main/java/com/scottlogic/PostSorter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply src/main/java/com/scottlogic/PostSorter.java:7]: Nice one :) I would probably call the function `sort` though. I would encourage nouns for naming classes and verbs for naming methods (you can think of ending the sentences "this class is a..." or "this method will...").

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+        switch (orderIn) {
+
+            case ASC -> Collections.sort(remainingPosts, Comparator.comparing(UserPost::getAuthor));
+
+            case DESC -> Collections.sort(remainingPosts, Collections.reverseOrder(comparing(UserPost::getAuthor)));
+        }
+
+        if (namesNull != null) {
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/AuthorPostSorter.java:34]: `namesNull` is never actually going to be null. You can check its size or use `.isEmpty()` instead. (Although we might not need this if we can get `nullsLast` working).

File: src/main/java/com/scottlogic/ContentPostSorter.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+            case ASC -> Collections.sort(inputList, Comparator.comparing(UserPost::getContents, (content1, content2) -> {
+                if (content1.length() == content2.length()) {
+                    return 0;
+                }
+                if (content1.length() < content2.length()) {
+                    return 1;
+                }
+                else
+                    return -1;
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/ContentPostSorter.java:22]: It might be nice to extract this into a private method.

File: src/main/java/com/scottlogic/ContentPostSorter.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        return inputList;
+    }
+}
+//movies.sort(Comparator.comparing(Movie::getStarred, (star1, star2) -> {
+//    if(star1 == star2){
+//         return 0;
+//    }
+//    return star1 ? -1 : 1;
+//}));
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/ContentPostSorter.java:37]: Commented out code can be removed.

File: src/main/java/com/scottlogic/Main.java [hunk 8]
```diff
@@ -21,7 +21,10 @@ public static void main(String[] args) {
                 "An example of a post \nwith lines breaks.", 3);
 
         List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3);
+        AuthorPostSorter aps = new AuthorPostSorter();
+        aps.sort(userPosts, SortOrder.ASC);
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/Main.java:25]: I think we were going to remove this?

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC),
+                "Hello World!", 2);
+    }
+
+
+
+
+
+    @Test
+    public void authorSorter_mixedValidNames_correctListReturnedAsc() {
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:26]: The first part of the test name is usually the name of the method being tested, i.e. `sort`. I also think the 'ascending' part would be better placed in the middle section, which describes the setup. Similarly for subsequent tests.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe Bloggs");
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:28]: Nice job including two posts with the same author. This could probably be a test in its own right though - I'm generally keen on testing "minimal" scenarios.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java (hunk 3 above)
Comment (pling-scottlogic) [reply src/test/java/com/scottlogic/AuthorPostSorterTest.java:91]: Other flavours of nothingness might include empty strings for authors, or where your entire post is null (not just the author field), or where your entire _list_ is null.

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]
I haven't looked too closely at all your tests, because I know you've had a session on unit testing since making this PR. Interested to see what you might change after that!

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        List<UserPost> namesNull = new ArrayList<>();
+        List<UserPost> remainingPosts = new ArrayList<>();
+
+        for (int i = 0; i < inputList.size(); i++) {
+            if (inputList.get(i).getAuthor() == null) {
+                namesNull.add(inputList.get(i));
+            } else {
+                remainingPosts.add(inputList.get(i));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/AuthorPostSorter.java:24]: No. I've not managed to get it working

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+
+        switch (orderIn) {
+
+            case ASC -> Collections.sort(remainingPosts, Comparator.comparing(UserPost::getAuthor));
```
[the remark this answers is not in the document]
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/AuthorPostSorter.java:29]: It's new from Java 12. There is also no need for a break statement with this enhanced switch.

File: src/main/java/com/scottlogic/Main.java (hunk 8 above)
Comment (amohanan-scottlogic) [reply src/main/java/com/scottlogic/Main.java:25]: I did.. don't know why its not reflected here. This is how I have it now "AuthorPostSorter sortByAuthor = new AuthorPostSorter();"
Then today, I've made it private as well

File: src/main/java/com/scottlogic/FullNamePostSorter.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+            case ASC -> Collections.sort(remainingPosts, Comparator.comparing(UserPost::getAuthor, (name1, name2) -> {
+                String firstName1;
+                String firstName2;
+                String lastName1;
+                String lastName2;
+
+                String[] split1 = name1.split(" ");
+                firstName1 = split1[0];
+                if (split1.length == 1) lastName1 = " ";
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/FullNamePostSorter.java:35]: I'm not keen of inline if statements like this. If you want to condense your control flow, have a look into the [ternary operator](https://www.baeldung.com/java-ternary-operator).

File: src/main/java/com/scottlogic/FullNamePostSorter.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            if (userPost.getAuthor() == null)
+                namesNull.add(userPost);
+            else if (userPost.getAuthor().isBlank())
+                namesSpace.add(userPost);
+            else
+                remainingPosts.add(userPost);
+        }
+        switch (orderIn) {
+
+            case ASC -> Collections.sort(remainingPosts, Comparator.comparing(UserPost::getAuthor, (name1, name2) -> {
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/FullNamePostSorter.java:27]: Lambda functions are generally good for small snippets of code. If you have this much code, maybe try refactoring it into a private method.

File: src/main/java/com/scottlogic/FullNamePostSorter.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+                String firstName1;
+                String firstName2;
+                String lastName1;
+                String lastName2;
+
+                String[] split1 = name1.split(" ");
+                firstName1 = split1[0];
+                if (split1.length == 1) lastName1 = " ";
+                else lastName1 = split1[1];
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/FullNamePostSorter.java:36]: What would happen if someone uses a middle name?

File: src/main/java/com/scottlogic/FullNamePostSorter.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+                String[] split2 = name2.split(" ");
+                firstName2 = split2[0];
+                if (split2.length == 1) lastName2 = " ";
+                else lastName2 = split2[1];
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/FullNamePostSorter.java:40]: There's a bit of repeated code here. It might be nice to introduce some more reusable private methods, like `getLastName()`.

File: src/main/java/com/scottlogic/FullNamePostSorter.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                    else if (firstName1.compareTo(firstName2) > 0)
+                        return 1;
+                    else return -1;
+
+                } else if (lastName1.compareTo(lastName2) > 0)
+                    return 1;
+                else return -1;
+            }));
+
+            case DESC -> Collections.sort(remainingPosts, Comparator.comparing(UserPost::getAuthor, (name1, name2) -> {
```
Comment (pling-scottlogic) [inline src/main/java/com/scottlogic/FullNamePostSorter.java:55]: Is there a way we can avoid repeating a lot of code for the descending case?

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        UserPost userPost3 = createUserPost("May Bloggs");
+        UserPost userPost4 = createUserPost("Joe Blogg");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4);
+        List<UserPost> sortByMethod = sortByAuthor.sort(userPosts, SortOrder.ASC);
+        List<UserPost> sorted = Arrays.asList(userPost4, userPost1, userPost2, userPost3);
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:30]: It might be clearer to call these something like `expectedResult` and `actualResult`.

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+        Assert.assertEquals(sortByMethod, sorted);
+    }
+
+    @Test
+    public void sort_nullNamesAsc_returnNullLast() {
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe Bloggs");
+        UserPost userPost3 = createUserPost("May Bloggs");
+        UserPost userPost4 = createUserPost("Joe Blogg");
+        UserPost userPost5 = createUserPost(null);
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:49]: We could probably reduce this to 2 or 3 input posts, since we're looking to verify a specific aspect of the functionality. Same for the test below.

File: src/test/java/com/scottlogic/ContentPostSorterTest.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        UserPost userPost3 = createUserPost("Hello World!");
+
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> sortByMethod = sortByContent.sort(userPosts, SortOrder.ASC);
+        List<UserPost> sorted = Arrays.asList(userPost3, userPost2, userPost1);
+        Assert.assertEquals(sortByMethod, sorted);
+    }
+
+    @Test
+    public void sort_sameLengthPostsAsc_correctListReturned() {
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/ContentPostSorterTest.java:35]: What is the expected outcome when there are two posts of the same length? Is the tie broken on another criteria, or does it maintain the input order? Is it any different when sorting in descending order?

File: src/test/java/com/scottlogic/ContentPostSorterTest.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+    public void sort_nullContentDesc_nullLastReturned() {
+        UserPost userPost1 = createUserPost(null);
+        UserPost userPost2 = createUserPost("An example of a post \nwith lines breaks.");
+        UserPost userPost3 = createUserPost("Hello World!");
+
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> sortByMethod = sortByContent.sort(userPosts, SortOrder.DESC);
+        List<UserPost> sorted = Arrays.asList(userPost2, userPost3, userPost1);
+        Assert.assertEquals(sortByMethod, sorted);
+    }
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/ContentPostSorterTest.java:118]: Can we include some tests for null posts and null lists?

File: src/test/java/com/scottlogic/FullNamePostSorterTest.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+    public void fullNameSorter_spaceForLastName_correctListReturnedDesc() {
+        UserPost userPost1 = createUserPost("Joe Bloggs");
+        UserPost userPost2 = createUserPost("Joe ");
+        UserPost userPost3 = createUserPost("May Bloggs");
+        UserPost userPost4 = createUserPost("Joe Bloggs");
+        List<UserPost> userPosts = Arrays.asList(userPost1, userPost2, userPost3, userPost4);
+        List<UserPost> sortByMethod = sortByFullName.sort(userPosts, SortOrder.DESC);
+        List<UserPost> sorted = Arrays.asList(userPost3, userPost1, userPost4, userPost2);
+        Assert.assertEquals(sortByMethod, sorted);
+    }
```
Comment (pling-scottlogic) [inline src/test/java/com/scottlogic/FullNamePostSorterTest.java:92]: Have a think about some more edge cases for this one. Middle names, nulls, empty strings...

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]
Nice work. Try and cover every scenario and edge case you can think of in the tests. In general, I would rather see a lot of small tests, each covering a specific scenario, than a smaller number of tests each with a lot of inputs.

File: src/test/java/com/scottlogic/ContentPostSorterTest.java (hunk 20 above)
Comment (amohanan-scottlogic) [reply src/test/java/com/scottlogic/ContentPostSorterTest.java:35]: I'm not sure if it hasn't updated on github. I've added testcases for this scenario. If the length is the same, it maintains input order for both ascending and descending.
